src/aria/grid/grid.ts

```diff
@@ -0,0 +1,272 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {_IdGenerator} from '@angular/cdk/a11y';
+import {
+  afterRenderEffect,
+  booleanAttribute,
+  computed,
+  contentChild,
+  contentChildren,
+  Directive,
+  ElementRef,
+  inject,
+  input,
+  model,
+  signal,
+  Signal,
+  untracked,
+} from '@angular/core';
+import {GridPattern, GridRowPattern, GridCellPattern, GridCellWidgetPattern} from '../ui-patterns';
+
+/** A directive that provides grid-based navigation and selection behavior. */
+@Directive({
+  selector: '[ngGrid]',
+  exportAs: 'ngGrid',
+  host: {
+    'class': 'grid',
+    'role': 'grid',
+    '[tabindex]': 'pattern.tabIndex()',
+    '[attr.aria-disabled]': 'pattern.disabled()',
+    '[attr.aria-activedescendant]': 'pattern.activeDescendant()',
+    '(keydown)': 'pattern.onKeydown($event)',
+    '(pointerdown)': 'pattern.onPointerdown($event)',
+    '(pointermove)': 'pattern.onPointermove($event)',
+    '(pointerup)': 'pattern.onPointerup($event)',
+    '(focusin)': 'onFocusIn()',
+    '(focusout)': 'onFocusOut()',
+  },
+})
+export class Grid {
+  /** The rows that make up the grid. */
+  private readonly _rows = contentChildren(GridRow);
+
+  /** The UI patterns for the rows in the grid. */
+  private readonly _rowPatterns: Signal<GridRowPattern[]> = computed(() =>
+    this._rows().map(r => r.pattern),
+  );
+
+  /** Whether selection is enabled for the grid. */
+  readonly enableSelection = input(false, {transform: booleanAttribute});
+
+  /** Whether the grid is disabled. */
+  readonly disabled = input(false, {transform: booleanAttribute});
+
+  /** Whether to skip disabled items during navigation. */
+  readonly skipDisabled = input(true, {transform: booleanAttribute});
+
+  /** The focus strategy used by the tree. */
+  readonly focusMode = input<'roving' | 'activedescendant'>('roving');
+
+  /** The wrapping behavior for keyboard navigation along the row axis. */
+  readonly rowWrap = input<'continuous' | 'loop' | 'nowrap'>('loop');
+
+  /** The wrapping behavior for keyboard navigation along the column axis. */
+  readonly colWrap = input<'continuous' | 'loop' | 'nowrap'>('loop');
+
+  /** The UI pattern for the grid. */
+  readonly pattern = new GridPattern({
+    ...this,
+    rows: this._rowPatterns,
+    getCell: e => this._getCell(e),
+  });
+
+  /** Whether the focus is in the grid. */
+  private readonly _hasFocus = signal(false);
+
+  constructor() {
+    afterRenderEffect(() => {
+      this.pattern.resetState();
```

Comment:
It's not immediately clear how often this function is running or how expensive this function is. Would it be possible to place a safeguard if statement around this call so that we are confident this won't rerun unnecessarily?